### ISSUE TITLE
Cookie, subtitle and playback problems

### ISSUE THREAD

**Issue by Goruti** [Varstahl/xbmc#12]
Ok, I finallygot some time to sit down and play a bit with this.

I got the following errors:

1.  When Trying to log in
```
20:47:55.403 T:1690391296   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.ValueError'>
                                            Error Contents: timestamp out of range for platform time_t
                                            Traceback (most recent call last):
                                              File "/home/pi/.kodi/addons/plugin.video.amazon-test/default.py", line 3021, in <module>
                                                exec mode + '()'
                                              File "<string>", line 1, in <module>
                                              File "/home/pi/.kodi/addons/plugin.video.amazon-test/default.py", line 2197, in LogIn
                                                cj.save(CookieFile, ignore_discard=True, ignore_expires=False)
                                              File "/usr/lib/python2.7/_LWPCookieJar.py", line 89, in save
                                                f.write(self.as_lwp_str(ignore_discard, ignore_expires))
                                              File "/usr/lib/python2.7/_LWPCookieJar.py", line 75, in as_lwp_str
                                                r.append("Set-Cookie3: %s" % lwp_cookie_str(cookie))
                                              File "/usr/lib/python2.7/_LWPCookieJar.py", line 35, in lwp_cookie_str
                                                time2isoz(float(cookie.expires))))
                                              File "/usr/lib/python2.7/cookielib.py", line 99, in time2isoz
                                                year, mon, mday, hour, min, sec = time.gmtime(t)[:6]
                                            ValueError: timestamp out of range for platform time_t
                                            -->End of Python script error report<--
```
Doing some research I found there is an know issue with LWPCookieJar (to handle 32/64 bits platforms). **To Fix it,** I replace LWPCookieJar by MozillaCookieJar.

1. Then I got a KEY error when it tries to load amzLang 
`amzLang = cj._cookies['.primevideo.com']['/']['lc-main-av'].value`
**To fix it,** I just modify for the value I saw in the cookies:
`amzLang = cj._cookies['.amazon.com']['/']['lc-main-av'].value`

1. I got a couple of 
> local variable 'xxxx' referenced before assignment
 in IStreamPlayback function

**To fix it,** I initialized these variables:
```
title = ""
mpaa_check = ""
thumb =  ""
```

1. I also have to disable the Subtitles because it was throwing me an error... which I did not debug further... 

Currently I am able to see all the content and navigate among all folders. But I am getting the following error when I try to play a video/TvShow.
```
22:32:17.730 T:1910763520  NOTICE: CVideoPlayer::CloseFile()
22:32:17.730 T:1910763520  NOTICE: VideoPlayer: waiting for threads to exit
22:32:17.730 T:1910763520  NOTICE: VideoPlayer: finished waiting
22:32:17.731 T:1910763520  NOTICE: CVideoPlayer::CloseFile()
22:32:17.731 T:1910763520  NOTICE: VideoPlayer: waiting for threads to exit
22:32:17.731 T:1910763520  NOTICE: VideoPlayer: finished waiting
22:32:27.558 T:1354281728  NOTICE: [Amazon VOD] Using Host: Limelight
22:32:27.563 T:1354281728  NOTICE: [Amazon VOD] checkURL: https://s3.ll.dash.row.aiv-cdn.net/d/2$0SofVIgO5tsknSqS7Z6cGxO84S0~/iad/3969/a893/b7bf/4d59-adb4-eb5b316f60f1/57b7cb37-78e9-4977-a1fc-e38d5a082661_corrected.mpd
22:32:27.779 T:1354281728  NOTICE: [Amazon VOD] getURL: https://s3.ll.dash.row.aiv-cdn.net/d/2$0SofVIgO5tsknSqS7Z6cGxO84S0/iad/3969/a893/b7bf/4d59-adb4-eb5b316f60f1/57b7cb37-78e9-4977-a1fc-e38d5a082661_corrected.mpd
22:32:29.052 T:1354281728  NOTICE: [Amazon VOD] Using inputstream.adaptive Version:2.0.10
22:32:29.162 T:1354281728  NOTICE: [Amazon VOD] Checking AudioTrack
22:32:29.205 T:1910763520  NOTICE: VideoPlayer: Opening: https://s3.ll.dash.row.aiv-cdn.net/d/2$0SofVIgO5tsknSqS7Z6cGxO84S0/iad/3969/a893/b7bf/4d59-adb4-eb5b316f60f1/57b7cb37-78e9-4977-a1fc-e38d5a082661_corrected.mpd
22:32:29.205 T:1910763520 WARNING: CDVDMessageQueue(player)::Put MSGQ_NOT_INITIALIZED
22:32:29.206 T:1708217088  NOTICE: Creating InputStream
22:32:33.357 T:1708217088   ERROR: AddOnLog: InputStream Adaptive: Unable to find license;hdcpEnforcementResolutionPixels in JSON string
22:32:33.357 T:1708217088   ERROR: AddOnLog: InputStream Adaptive: License update not successful (no keys)
22:32:33.358 T:1708217088   ERROR: AddOnLog: InputStream Adaptive: Initialize failed (SingleSampleDecrypter)
22:32:33.360 T:1708217088   ERROR: CVideoPlayer::OpenInputStream - error opening [https://s3.ll.dash.row.aiv-cdn.net/d/2$0SofVIgO5tsknSqS7Z6cGxO84S0/iad/3969/a893/b7bf/4d59-adb4-eb5b316f60f1/57b7cb37-78e9-4977-a1fc-e38d5a082661_corrected.mpd]
22:32:33.360 T:1708217088  NOTICE: CVideoPlayer::OnExit()
22:32:33.360 T:1910763520   ERROR: Playlist Player: skipping unplayable item: 0, path [plugin://plugin.video.amazon-test/?mode=PrimeVideo_Browse&path=root-%2F%2F-TV+Shows-%2F%2F-Action+%26+Adventure+TV-%2F%2F-amzn1.dv.gti.28abcdd1-c613-cfef-f043-5fac26c5ab27-%2F%2F-0FFL11Y1WDDFRM8ETA0KLJEEM3-//-0I9VW5W7RJFXEN80MZKX9UEJE3]
22:32:33.372 T:1910763520  NOTICE: CVideoPlayer::CloseFile()
22:32:33.372 T:1910763520  NOTICE: VideoPlayer: waiting for threads to exit
22:32:33.372 T:1910763520  NOTICE: VideoPlayer: finished waiting
22:32:33.373 T:1910763520  NOTICE: CVideoPlayer::CloseFile()
22:32:33.373 T:1910763520  NOTICE: VideoPlayer: waiting for threads to exit
22:32:33.373 T:1910763520  NOTICE: VideoPlayer: finished waiting
```

Any idea what this error means? And how we can fix it?

Thanks you very much!

**Comment by Varstahl** [comment]
Some widevine problems, I'll look into that. Since cookielib is starting to cause problems, I'd say let's remove external cookiejars and use the standard requests one.

**Comment by Varstahl** [comment]
Update: the subtitles problem was due to the change to requests. I've fixed it together with the audio streams. I'll eat the cookie problem later today.

**Comment by Varstahl** [comment]
Removed the LWPCookieJar dependency in favour of Requests' embedded system, fixed in b7f2373b5a4bc863a3b528964eab9213f8ddb556.

**Comment by Varstahl** [comment]
@Goruti I believe the license problem was due to cookies, if with the last update the problem isn't fixed feel free to reopen the ticket and I'll try to address it.

**Comment by Goruti** [comment]
Thanks for looking at it @Varstahl.

I tried the latest version and still some InputStream issue :(

I loaded DEBUG logs [here](http://paste.debian.net/1007525/)

**Comment by Varstahl** [comment]
Could it be a duplicate of https://github.com/Sandmann79/xbmc/issues/62?

I also see you're using `~/.kodi/cdm/libssd_wv.so`, while the (only?) known working widevine library is the `libwidevinecdm.so` by Google.

I don't have a rasp around, I might try with an orange at a later date, but there are a few guides like [this one](https://gist.github.com/ruario/3c873d43eb20553d5014bd4d29fe37f1) that can maybe help you changing the decoding library.

**Comment by Goruti** [comment]
No luck. It seems that the new widevine library that is required by AmazonPrime does not work on RPI yet :(

I am currently running other Addon (Dazn) which requires widevine and it works fine with my current configuration (**~/.kodi/cdm/libssd_wv.so** and **~/.kodi/cdm/libwidevinecdm.so**)

**Comment by Varstahl** [comment]
Yeah, there have been some updates to widevine DRM, that required some work on several fronts.

I just re-checked the script, and since RaspberryPI is arm based that script won't work since it download 32/64bit versions of chrome, but not the ARM (if there's any).

Question: are you able to install Chrome on your rasp and access primevideo.com through it? If you can you can just then copy the libwidevinecdm.so that Chrome installs over to your kodi/cdm directory, and it should work. I haven't got my hands back on the OrangePI yet, so I can't test.

**Comment by CGarces** [comment]
Hi @Goruti, can you please try this?

Open a ssh over your LibreELEC instalation and execute this:

curl -Ls http://nmacleod.com/public/libreelec/getwidevine.sh | bash

**Comment by Sandmann79** [comment]
@CGarces this script doesn't work anymore, because it downloads an outdated version (1.4.8.823) of libwidevinecdm.so, which is actually rejected by Amazon. 
The inputstream.helper, which is integrated since last version update, downloads the current widevine lib. But this one decodes the YUV Pictures in another way then before. The Kodi internal OpenGL/-ES render can handle this, that's the reason why it works on other devices. But RPi uses another rendering engine, this one should do this too, but it needs some time for development and a new firmware.

**Comment by Varstahl** [comment]
@Sandmann79 wasn't aware of that, thanks for the information.

**Comment by Goruti** [comment]
@Varstahl Yes I installed a [custom version of chromium](https://thepi.io/how-to-watch-netflix-on-the-raspberry-pi/) on my pi which allowed (or used to...) to me watch netflix and Amazon Prime (that's how I got widevine).

Today I try to load any video using chromium (as i used to do) and it DID NOT work! :( It's asking me for a newest version of chromium. So yes, It seems that the problem is because RPI does not have the latest widevine code yet...
